webapp/api_handlers.go

```diff
@@ -374,3 +377,122 @@ func handleAPIDiffPost(w http.ResponseWriter, r *http.Request) {
 	}
 	w.Write(bytes)
 }
+
+func apiManifestHandler(w http.ResponseWriter, r *http.Request) {
+	sha, err := ParseSHAParamFull(r)
+	if err != nil {
+		http.Error(w, err.Error(), http.StatusBadRequest)
+		return
+	}
+	ctx := appengine.NewContext(r)
+	if manifest, err := getManifestForSHA(ctx, sha); err != nil {
+		http.Error(w, err.Error(), http.StatusNotFound)
+	} else if manifest == nil {
+		http.NotFound(w, r)
+	} else {
+		w.Header().Add("content-type", "application/json")
+		w.Write(manifest)
+	}
+}
+
+func getManifestForSHA(ctx context.Context, sha string) (manifest []byte, err error) {
+	// Fetch models.Token entity for GitHub API Token.
+	tokenKey := datastore.NewKey(ctx, "Token", "github-api-token", 0, nil)
+	var token models.Token
+	datastore.Get(ctx, tokenKey, &token)
+
+	// Search for the PR associated with the SHA.
+	const githubSearch = `https://api.github.com/search/issues?q=SHA:%s+user:w3c+repo:web-platform-tests`
+	url := fmt.Sprintf(githubSearch, sha)
+	var body []byte
+	if body, err = fetchGitHubURL(ctx, url, token.Secret); err != nil {
+		return nil, err
+	}
+
+	var queryResults map[string]*json.RawMessage
+	if err = json.Unmarshal(body, &queryResults); err != nil {
+		return nil, err
+	}
+	var issues []map[string]*json.RawMessage
+	if err = json.Unmarshal(*queryResults["items"], &issues); err != nil {
+		return nil, err
+	}
+	if len(issues) < 1 {
+		return nil, fmt.Errorf("No search results found for SHA %s", sha)
+	}
+
+	// Load the release by the presumed tag name merge_pr_*
+	var prNumber int
+	if err = json.Unmarshal(*issues[0]["number"], &prNumber); err != nil {
+		return nil, err
+	}
+
+	releaseTag := fmt.Sprintf("merge_pr_%d", prNumber)
+	const githubRelease = `https://api.github.com/repos/w3c/web-platform-tests/releases/tags/%s`
+	url = fmt.Sprintf(githubRelease, releaseTag)
+	if body, err = fetchGitHubURL(ctx, url, token.Secret); err != nil {
+		return nil, err
+	}
+
+	var release map[string]*json.RawMessage
+	if err = json.Unmarshal(body, &release); err != nil {
+		return nil, err
+	}
+	var assets []map[string]*json.RawMessage
+	if err = json.Unmarshal(*release["assets"], &assets); err != nil {
+		return nil, err
+	}
+	if len(assets) < 1 {
+		return nil, fmt.Errorf("No assets found for release %s", releaseTag)
+	}
+	// Get (and unzip) the asset with name "MANIFEST-{sha}.json.gz"
```

Comment:
Do we need to unzip it, or could we set the transfer encoding and just pass along the data?

Reply:
We played with this a bit. The short answer is that we can't, unless (potentially) we use some dirty hacks.

Basically, GitHub responds a 302 for the download URL to a long AWS S3 URL that has a bunch of query parameters to tell S3 to force a browser download, which means S3 will disregard the `Accept-Encoding` in request headers, and always return the raw octet stream of the file (which happens to be a gz here).

If we absolutely want to to do it, I guess we can intercept 302 and modify the query parameters, but I guess we don't :)

Reply:
We want to unzip it anyway, once we get to dumping it in the cache and allowing path filtering, etc.